### PR TITLE
Dfview

### PR DESCRIPTION
**Pull request by deng113jie**
df.view() to create a view from the df, contains all fields as a reference
df.apply_filter() to set filter

### CODE REVIEW

File: exetera/core/dataframe.py [hunk 1]
```diff
@@ -478,19 +596,25 @@ def apply_filter(self, filter_to_apply, ddf=None):
...
                 raise TypeError("The destination object must be an instance of DataFrame.")
+            ddf._write_filter(np.where(filter_to_apply_ == True)[0])
             for name, field in self._columns.items():
-                newfld = field.create_like(ddf, name)
-                field.apply_filter(filter_to_apply_, target=newfld)
+                # hard copy
+                # newfld = field.create_like(ddf, name)
+                # field.apply_filter(filter_to_apply_, target=newfld)
+                # soft copy - view
+                newfld = ddf.add_view(field)
```
Comment (deng113jie) [inline exetera/core/dataframe.py:608]: _add_view

Comment (deng113jie) [reply exetera/core/dataframe.py:608]: _add_view with filter (passing the filter in), remove the filter setter

File: exetera/core/fields.py [hunk 2]
```diff
@@ -2564,6 +2591,54 @@ def data(self):
...
+        self._value_wrapper = FieldArray
+        if self.is_view():
+            self.data.register_reference(self)
+
+    @property
+    def filter(self):
+        if self._filter_wrapper is None:
+            return None
+        else:
+            return self._filter_wrapper[:]
```
Comment (deng113jie) [inline exetera/core/fields.py:2608]: just return the wrapper, not data

File: exetera/core/fields.py [hunk 3]
```diff
@@ -340,6 +365,8 @@ def write(self, part):
         :param part: numpy array to write to field
         :return: None
         """
+        if len(self._references) > 0:
+            self.concreate_all_fields()
```
Comment (deng113jie) [inline exetera/core/fields.py:369]: move these into field, so that the callback flow is: fieldarray -> field -> field_references

File: exetera/core/fields.py [hunk 4]
```diff
@@ -145,7 +154,14 @@ def _ensure_valid(self):
         if not self._valid_reference:
             raise ValueError("This field no longer refers to a valid underlying field object")
 
-
+    def is_view(self):
+        """
+        Return if the dataframe's name matches the field h5group path; if not, means this field is a view.
+        """
+        if self._field.name[1:1+len(self.dataframe.name)] != self.dataframe.name:
```
Comment (ericspod) [inline exetera/core/fields.py:161]:
```suggestion
        return self._field.name[1:1+len(self.dataframe.name)] != self.dataframe.name
```

File: exetera/core/fields.py [hunk 5]
```diff
@@ -243,7 +264,11 @@ def dtype(self):
         return self._dataset.dtype
 
     def __getitem__(self, item):
-        return self._dataset[item]
+        if self._field_instance.filter is not None:
+            mask = self._field_instance.filter[:]
```
Comment (ericspod) [inline exetera/core/fields.py:268]: Apply the item to select indices from the mask first then use these to get values from self._dataset:
```
idx = mask[item]
```

File: exetera/core/fields.py [hunk 6]
```diff
@@ -303,27 +328,13 @@ def dtype(self):
...
-
-        """
-        for field in self._references:
-            field.concrete_reference()
-
     def __getitem__(self, item):
-        return self._dataset[item]
+        if self._field_instance.filter is not None and not isinstance(self._field_instance, IndexedStringField):
+            mask = self._field_instance.filter[:]
+            data = self._dataset[:][mask]  # as HDF5 does not support unordered mask
```
Comment (ericspod) [inline exetera/core/fields.py:334]: Same idea here, want to get indices from mask rather than resolve all mask indices then select from result.

File: exetera/core/fields.py [hunk 7]
```diff
@@ -483,16 +494,17 @@ def complete(self):
 
 
 class ReadOnlyIndexedFieldArray:
-    def __init__(self, field, indices, values):
+    def __init__(self, chunksize, indices, values, field):
```
Comment (ericspod) [inline exetera/core/fields.py:497]: TODO: check this is right?

File: exetera/core/fields.py [hunk 8]
```diff
@@ -522,27 +534,48 @@ def __getitem__(self, item):
...
+            if self._field_instance.filter is None:
+                if item >= len(self._indices) - 1:
+                    raise ValueError("index is out of range")
+                start, stop = self._indices[item:item + 2]
+                if start == stop:
+                    return ''
+                value = self._values[start:stop].tobytes().decode()
+                return value
+            else:
+                mask = self._field_instance.filter[:]
```
Comment (ericspod) [inline exetera/core/fields.py:570]: Samesies.

File: exetera/core/dataframe.py [hunk 9]
```diff
@@ -478,13 +524,12 @@ def apply_filter(self, filter_to_apply, ddf=None):
...
-
-        if ddf is not None:
+        if ddf is not None and ddf is not self:
             if not isinstance(ddf, DataFrame):
                 raise TypeError("The destination object must be an instance of DataFrame.")
+            filter_to_apply_ = filter_to_apply_.nonzero()[0]
             for name, field in self._columns.items():
-                newfld = field.create_like(ddf, name)
-                field.apply_filter(filter_to_apply_, target=newfld)
+                ddf._add_view(field, filter_to_apply_)
```
Comment (deng113jie) [inline exetera/core/dataframe.py:532]: check if the same dataset

File: exetera/core/fields.py [hunk 10]
```diff
@@ -2922,11 +3096,36 @@ def data(self):
...
+            self._view_refs.clear()
+        else:
+            self._view_refs.remove(view)
+
+    def notify(self, msg=None):
+        for view in self._view_refs:
+            view.update(self, msg)
+
+    def update(self, subject, msg=None):
+        if isinstance(subject, (WriteableFieldArray, WriteableIndexedFieldArray)):
```
Comment (atbenmurray) [inline exetera/core/fields.py:3118]: `# This field is being notified by its own field array`
`# It needs to notify other fields that it is about to change before the change goes ahead`

File: exetera/core/fields.py [hunk 11]
```diff
@@ -2922,11 +3096,36 @@ def data(self):
...
+            self.notify(msg)
+            self.detach()
+
+        if isinstance(subject, HDF5Field):
```
Comment (atbenmurray) [inline exetera/core/fields.py:3122]: `# This field is being notified by the field that owns the data that it has a view of`
`# At present, the behavior is that it copies the data and then detaches from the view that notified it, as it`
`# no longer has an observation relationship with that field`

File: exetera/core/fields.py [hunk 12]
```diff
@@ -2922,11 +3096,36 @@ def data(self):
...
+            self._view_refs.remove(view)
+
+    def notify(self, msg=None):
+        for view in self._view_refs:
+            view.update(self, msg)
+
+    def update(self, subject, msg=None):
+        if isinstance(subject, (WriteableFieldArray, WriteableIndexedFieldArray)):
+            self.notify(msg)
+            self.detach()
```
Comment (atbenmurray) [inline exetera/core/fields.py:3120]: Detach should be the responsibility of the observer, not the subject, as the subject could instead do something clever that maintains the relationship

File: exetera/core/fields.py [hunk 13]
```diff
@@ -2922,11 +3096,36 @@ def data(self):
...
+                self._value_wrapper = WriteableFieldArray(self, 'values')
             else:
-                self._value_wrapper = ReadOnlyFieldArray(self._field, 'values')
+                self._value_wrapper = ReadOnlyFieldArray(self, 'values')
         return self._value_wrapper
 
+    def attach(self, view):
+        self._view_refs.append(view)
+
+    def detach(self, view=None):
```
Comment (atbenmurray) [inline exetera/core/fields.py:3107]: This detach is actually `notify_deletion`. This is a standard part of subject_observer, but so is `detach`, which is for the observer to detach from a given subject

File: exetera/core/fields.py [hunk 14]
```diff
@@ -2104,9 +2203,34 @@ def data(self):
...
             wrapper = \
                 WriteableIndexedFieldArray if self._write_enabled else ReadOnlyIndexedFieldArray
-            self._data_wrapper = wrapper(self.chunksize, self.indices, self.values)
+            self._data_wrapper = wrapper(self.chunksize, self.indices, self.values, self)
         return self._data_wrapper
 
+    def attach(self, view):
+        self._view_refs.append(view)
+
+    def detach(self, view=None):
```
Comment (deng113jie) [inline exetera/core/fields.py:2212]: also deletion function (subject del observer)

File: exetera/core/fields.py [hunk 15]
```diff
@@ -300,7 +330,14 @@ def dtype(self):
         return self._dataset.dtype
 
     def __getitem__(self, item):
-        return self._dataset[item]
+        if self._field_instance.filter is not None and not isinstance(self._field_instance, IndexedStringField):
+            mask = self._field_instance.filter[item]
+            if utils.is_sorted(mask):
+                return self._dataset[mask]
+            else:
+                return self._dataset[np.sort(mask)][mask]
```
Comment (deng113jie) [inline exetera/core/fields.py:338]: add an issue on benchmarking: filtering/item -> hdf5 or hdf5/numpy -> filter.

File: exetera/core/fields.py [hunk 16]
```diff
@@ -495,27 +537,55 @@ def __getitem__(self, item):
...
-            index = self._indices[start:stop + 1]
-            bytestr = self._values[index[0]:index[-1]]
-            results = [None] * (len(index) - 1)
-            startindex = self._indices[start]
-            for ir in range(len(results)):
-                results[ir] = \
-                    bytestr[index[ir] - np.int64(startindex):
-                            index[ir + 1] - np.int64(startindex)].tobytes().decode()
-            return results
+            if self._field_instance.filter is None:
```
Comment (atbenmurray) [inline exetera/core/fields.py:540]: `# This field is not a view so no filtered_index to deal with`

File: exetera/core/fields.py [hunk 17]
```diff
@@ -495,27 +537,55 @@ def __getitem__(self, item):
...
+                startindex = self._indices[start]
+                for ir in range(len(results)):
+                    results[ir] = \
+                        bytestr[index[ir] - np.int64(startindex):
+                                index[ir + 1] - np.int64(startindex)].tobytes().decode()
+                return results
+            else:
+                mask = self._field_instance.filter[item]
+                if utils.is_sorted(mask):
+                    index_s = self._indices[mask]
```
Comment (atbenmurray) [inline exetera/core/fields.py:557]: `# the filtered indices represent a filter operation`
We need to evaluate whether this saves anything. H5py is horrifically slow and the risk is that all of this work is lost when loading slices through its api. I would suggest just loading all of the data and applying the filtered index rather than trying to gain time here, unless we do a series of detailed benchmarks that can give us a heuristic to decide whether to do this or not

File: exetera/core/dataframe.py [hunk 18]
```diff
@@ -478,13 +524,12 @@ def apply_filter(self, filter_to_apply, ddf=None):
         :returns: a dataframe contains all the fields filterd, self if ddf is not set
         """
         filter_to_apply_ = val.validate_filter(filter_to_apply)
-
-        if ddf is not None:
+        if ddf is not None and ddf is not self:
```
[the remark this answers is not in the document]
Comment (ericspod) [reply exetera/core/dataframe.py]: `if ddf not in (None, self)`
